asserts/database.go

```diff
@@ -176,3 +223,101 @@ func (db *Database) Check(assert Assertion) error {
 	}
 	return fmt.Errorf("failed signature verification: %v", lastErr)
 }
+
+// Add persists the assertions after ensuring it is properly signed and consistent with all the stored knowledge.
+// Returns ErrAssertionNotSuperseding if trying to add a previous revision of the assertion.
+func (db *Database) Add(assert Assertion) error {
+	reg, err := checkAssertType(assert.Type())
+	if err != nil {
+		return err
+	}
+	err = db.Check(assert)
+	if err != nil {
+		return err
+	}
+	primaryKey := make([]string, len(reg.primaryKey))
+	for i, k := range reg.primaryKey {
+		keyVal := assert.Header(k)
+		if keyVal == "" {
+			return fmt.Errorf("missing primary key header: %v", k)
+		}
+		// safety against '/' etc
+		primaryKey[i] = url.QueryEscape(keyVal)
+	}
+	indexPath := filepath.Join(string(assert.Type()), filepath.Join(primaryKey...))
+	_, err = db.statEntry(assertionsRoot, indexPath)
+	switch {
+	case err == nil:
+		// directory for assertion is present
+		curRevStr, err := db.readlinkEntry(assertionsRoot, indexPath, "latest")
+		if err != nil {
+			return fmt.Errorf("broken assertion store, reading 'latest' revision symlink: %v", err)
+		}
+		curRev, err := strconv.Atoi(curRevStr)
+		if err != nil {
+			return fmt.Errorf("broken assertion store, extracting revision from 'latest' revision symlink: %v", err)
+		}
+		if curRev >= assert.Revision() {
+			return ErrAssertionNotSuperseding
+		}
+		err = db.removeEntry(assertionsRoot, indexPath, "latest")
+		if err != nil {
+			return fmt.Errorf("broken assertion store, could not remove 'latest' revision symlink: %v", err)
+		}
+	case os.IsNotExist(err):
+		// nothing there yet
+	default:
+		return fmt.Errorf("broken assertion store, failed to stat assertion directory: %v", err)
+	}
+	revStr := strconv.Itoa(assert.Revision())
+	err = db.atomicWriteEntry(Encode(assert), false, assertionsRoot, indexPath, revStr)
+	if err != nil {
+		return fmt.Errorf("broken assertion store, failed to write assertion: %v", err)
+	}
+	err = db.symlinkEntry(revStr, assertionsRoot, indexPath, "latest")
+	if err != nil {
+		return fmt.Errorf("broken assertion store, failed to create 'latest' revision symlink: %v", err)
+	}
+	return nil
+}
+
+// Find an assertion based on arbitrary headers.
+// Provided headers must contain the primary key for the assertion type.
+// Returns ErrAssertionNotFound if the assertion cannot be found.
+func (db *Database) Find(assertionType AssertionType, headers map[string]string) (Assertion, error) {
+	reg, err := checkAssertType(assertionType)
+	if err != nil {
+		return nil, err
+	}
+	primaryKey := make([]string, len(reg.primaryKey))
+	for i, k := range reg.primaryKey {
+		keyVal := headers[k]
+		if keyVal == "" {
+			return nil, fmt.Errorf("must provide primary key: %v", k)
+		}
+		primaryKey[i] = url.QueryEscape(keyVal)
+	}
+	indexPath := filepath.Join(string(assertionType), filepath.Join(primaryKey...))
+	_, err = db.statEntry(assertionsRoot, indexPath)
+	if err != nil {
+		if os.IsNotExist(err) {
```

Comment:
This can be the first thing after statEntry, to avoid the double nesting (it can take nil).